projects/chaos/src/sequencer_fsm.c

```diff
@@ -152,47 +206,80 @@ FSM_STATE_TRANSITION(sequencer_state_idle) {
 
 FSM_STATE_TRANSITION(sequencer_state_charge) {
   FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_EMERGENCY, sequencer_state_emergency);
+  FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_DRIVE, sequencer_state_drive);
   FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_IDLE, sequencer_state_idle);
   FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_RESET, sequencer_state_charge);
 }
 
 FSM_STATE_TRANSITION(sequencer_state_drive) {
   FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_EMERGENCY, sequencer_state_emergency);
+  FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_CHARGE, sequencer_state_charge);
   FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_IDLE, sequencer_state_idle);
   FSM_ADD_TRANSITION(CHAOS_EVENT_SEQUENCE_RESET, sequencer_state_drive);
 }
 
+// Return true if setup succeeded. Return false if we are blocked waiting on an async event to
+// finish. If this returns false the caller should exit immediately.
+static bool prv_sequencer_setup_common(void) {
+  if (sequencer_stop_awaiting(&s_storage)) {
+    relay_retry_service_fail_fast();
+    // We actually transitioned so just reset the state until it succeeds. Note this needs to be
+    // normal priority to allow events to flush. This is to flush an event we are waiting for
+    // (i.e. relay transition) this will take a  mas of 15 ms to succeed.
+    event_raise(CHAOS_EVENT_SEQUENCE_RESET, SEQUENCER_EMPTY_DATA);
+    return false;
+  }
+  s_retries = 0;
+  prv_clear_filters();
+  return true;
+}
+
 // FSM Transitions
 static void prv_sequencer_state_emergency(FSM *fsm, const Event *e, void *context) {
   (void)fsm;
   (void)e;
+  if (!prv_sequencer_setup_common()) {
+    return;
+  }
+  prv_add_filter(CHAOS_EVENT_RELAY_ERROR);
```

Comment:
What does filtering do?

Reply:
Normally the sequencer filters to the range of `(NUM_CHAOS_EVENTS_CAN, NUM_CHAOS_EVENT_SEQEUNCES)` with special handling of `CHAOS_EVENT_RELAY_ERROR`

We use filtering in the emergency state to discard `CHAOS_EVENT_RELAY_ERROR` and `CHAOS_EVENT_RELAY_OPENED` so that we can skip awaiting responses. However, if we didn't include this filtering behavior the `CHAOS_EVENT_RELAY_OPENED` message would hit the main sequencer and since it was unexpected it would cause a fault. To avoid this we need to filter it.

Reply:
Ah, can you document that?

Reply:
It is documented... But I will make it more explicit.